### PR TITLE
[Core] Adding scientific precission to ModelPartIO (transition #8576)

### PR DESCRIPTION
**Pull request by loumalouomega**
**Description**
Adding scientific precission to ModelPartIO (transition #8576)

**Changelog**
- Adding scientific precission to ModelPartIO (transition #8576)
- Some clean up
- Update mdpa tests

### CODE REVIEW

**Review by philbucher — CHANGES_REQUESTED** [review]
why is this necessary / was it discussed before?

This increases the file size! I played around and tested with this a lot in the past, this has a significant impact on the file size!

Maybe for small models (in terms of dimension) it can be beneficial/necessary but it should not be the default, maybe it can be optional

other opinions?

**Comment by loumalouomega** [comment]
It was requested and added in #8576, I separate it in here

**Comment by loumalouomega** [comment]
Requested by @vgudzulic

**Comment by philbucher** [comment]
ah ok
if it is needed then it can be made optional, but off by default

File: kratos/sources/model_part_io.cpp [hunk 1]
```diff
@@ -138,10 +131,11 @@ std::size_t ModelPartIO::ReadNodesNumber()
 
 void ModelPartIO::WriteNodes(NodesContainerType const& rThisNodes)
 {
-    (*mpStream) << "Begin Nodes" << std::endl;
-    for(NodesContainerType::const_iterator it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; it_node++)
-        (*mpStream) << "\t" << it_node->Id() << "\t" << it_node->X()  << "\t" << it_node->Y() << "\t" << it_node->Z() << std::endl;
-    (*mpStream) << "End Nodes" << std::endl << std::endl;
+    mpStream->precision(10);
+    (*mpStream) << "Begin Nodes\n" << std::scientific;
+    for(auto it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; it_node++)
```
Comment (philbucher) [inline kratos/sources/model_part_io.cpp]:
```suggestion
    for(NodesContainerType::const_iterator it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; ++it_node)
```

I am pretty sure that `auto` will result in a regular and not a const iterator!

File: kratos/sources/model_part_io.cpp [hunk 2]
```diff
@@ -138,10 +131,18 @@ std::size_t ModelPartIO::ReadNodesNumber()
...
+        (*mpStream) << "Begin Nodes" << std::endl;
+        for(NodesContainerType::const_iterator it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; it_node++)
+            (*mpStream) << "\t" << it_node->Id() << "\t" << it_node->X()  << "\t" << it_node->Y() << "\t" << it_node->Z() << "\n";
+    } else {
+        mpStream->precision(10);
+        (*mpStream) << "Begin Nodes\n" << std::scientific;
+        for(NodesContainerType::const_iterator it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; it_node++)
+            (*mpStream) << "\t" << it_node->Id() << "\t" << it_node->X()  << "\t" << it_node->Y() << "\t" << it_node->Z() << "\n";
+    }
+    (*mpStream) << "End Nodes\n\n";
```
Comment (philbucher) [inline kratos/sources/model_part_io.cpp]:
```suggestion
    // Printing or not with scientific precision
    if (mOptions.Is(IO::SCIENTIFIC_PRECISION)) {
        (*mpStream) << std::setprecision(10) << std::scientific;
    }
    (*mpStream) << "Begin Nodes" << std::endl;
    for(NodesContainerType::const_iterator it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; ++it_node)
        (*mpStream) << "\t" << it_node->Id() << "\t" << it_node->X()  << "\t" << it_node->Y() << "\t" << it_node->Z() << "\n";
    (*mpStream) << "End Nodes\n\n";
```

**Review by philbucher — COMMENTED** [review]
ok for me know, lets leave it open for some days to hear some more opinions

**Comment by loumalouomega** [comment]
I reverted one test

**Comment by philbucher** [comment]
@rubenzorrilla @roigcarlo what do you think about this PR?
I think we can merge it

**Review by roigcarlo — COMMENTED** [review]
My only concern was this not being able to be switchable, since it is I have no problem :+1:

**Comment by vgudzulic** [comment]
Thanks for this! Optional is fine. @philbucher @roigcarlo to clarify: this change is only in serial. In mpi parallel, writing .mdpa files is already done with more precision (10 decimal places, I believe). I have asked for this enhancement because I perform some changes in the .mdpa file in serial, write it out, and run the simulation parallel. In this process, I was losing a lot of precision because the decimal places were getting truncated. I believe that this serial function is typically not used.

**Comment by loumalouomega** [comment]
Ping @rubenzorrilla

**Comment by loumalouomega** [comment]
@rubenzorrilla you are the one missing

**Comment by loumalouomega** [comment]
@rubenzorrilla Check this please!

File: kratos/sources/model_part_io.cpp [hunk 3]
```diff
@@ -138,10 +131,14 @@ std::size_t ModelPartIO::ReadNodesNumber()
...
 void ModelPartIO::WriteNodes(NodesContainerType const& rThisNodes)
 {
+    // Printing or not with scientific precision
+    if (mOptions.Is(IO::SCIENTIFIC_PRECISION)) {
+        (*mpStream) << std::setprecision(10) << std::scientific;
+    }
     (*mpStream) << "Begin Nodes" << std::endl;
-    for(NodesContainerType::const_iterator it_node = rThisNodes.begin() ; it_node != rThisNodes.end() ; it_node++)
-        (*mpStream) << "\t" << it_node->Id() << "\t" << it_node->X()  << "\t" << it_node->Y() << "\t" << it_node->Z() << std::endl;
-    (*mpStream) << "End Nodes" << std::endl << std::endl;
```
Comment (pooyan-dadvand) [inline kratos/sources/model_part_io.cpp:144]: I would not change it to the ascii return

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee agrees with this change. Thanks for the implementation!

File: kratos/sources/model_part_io.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/sources/model_part_io.cpp:144]: Okidoki

Comment (loumalouomega) [reply kratos/sources/model_part_io.cpp:144]: (a priori it is better https://www.educative.io/edpresso/what-is-the-difference-between-endl-and-n-in-cpp)

Comment (loumalouomega) [reply kratos/sources/model_part_io.cpp:144]: Done

**Comment by philbucher** [comment]
guys using `std::endl` is much slower than `\n` since it flushes the stream each time!
See #7922

I also saw Jason Turner talk abt this in one of his videos, hence I strongly recommend against using `std::endl` inside the loop

File: kratos/sources/model_part_io.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/sources/model_part_io.cpp:144]: I strongly recommend using `\n`
See #7922

**Comment by loumalouomega** [comment]
> guys using `std::endl` is much slower than `\n` since it flushes the stream each time! See #7922
> 
> I also saw Jason Turner talk abt this in one of his videos, hence I strongly recommend against using `std::endl` inside the loop

Do I revert then?

File: kratos/sources/model_part_io.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply kratos/sources/model_part_io.cpp:144]: I think that @pooyan-dadvand referred to the last two outside the loop.

Comment (loumalouomega) [reply kratos/sources/model_part_io.cpp:144]: okay, done

**Review by philbucher — APPROVED** [review]
👍

**Comment by loumalouomega** [comment]
> +1

morpheus-at-last

**Pull request merged**